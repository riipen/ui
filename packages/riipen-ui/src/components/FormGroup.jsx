import clsx from "clsx";
import css from "styled-jsx/css";

import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import Typography from "./Typography";

class FormGroup extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * A title for the field set.
     */
    title: PropTypes.node,

    /**
     * A subtitle for the field set.
     */
    subtitle: PropTypes.node
  };

  static defaultProps = {
    classes: []
  };

  getLinkedStyles = () => {
    const theme = this.context;

    return css.resolve`
      .subtitle {
        color: ${theme.palette.text.secondary};
        font-weight: ${theme.typography.fontWeight.regular};
      }

      .title {
        font-weight: ${theme.typography.fontWeight.medium};
        font-size: 22px;
        margin-bottom: ${theme.spacing(3)}px;
      }
    `;
  };

  static contextType = ThemeContext;

  render() {
    const { children, classes, title, subtitle } = this.props;

    const theme = this.context;

    const linkedStyles = this.getLinkedStyles();
    const className = clsx(classes);

    return (
      <React.Fragment>
        <fieldset className={className}>
          {title && (
            <Typography
              classes={[linkedStyles.className, "title"]}
              component="p"
              variant="h3"
            >
              {title}
              {linkedStyles.styles}
            </Typography>
          )}
          {subtitle && (
            <Typography
              classes={[linkedStyles.className, "subtitle"]}
              component="p"
              gutter
              variant="h4"
            >
              {subtitle}
              {linkedStyles.styles}
            </Typography>
          )}
          <div className="children">{children}</div>
        </fieldset>
        <style jsx>{`
          fieldset {
            border: none;
            margin: 0 0 ${theme.spacing(8)}px 0;
            padding: 0;
          }

          fieldset:last-child {
            margin-bottom: 0;
          }

          .children {
            margin-top: ${theme.spacing(4)}px;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses(FormGroup);
