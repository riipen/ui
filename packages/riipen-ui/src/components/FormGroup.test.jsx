import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import FormGroup from "./FormGroup";

describe("<FormGroup>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<FormGroup />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<FormGroup />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <div>child</div>;

      const wrapper = mount(<FormGroup>{child}</FormGroup>);

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classes = ["newClass"];

      const wrapper = mount(<FormGroup classes={classes} />);

      expect(
        wrapper
          .find("FormGroup")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-formgroup"];

      const wrapper = mount(<FormGroup />);

      expect(wrapper.find("FormGroup").props().classes).toEqual(classes.sort());
    });
  });

  describe("title prop", () => {
    it("renders given title", () => {
      const title = "Title";

      const wrapper = mount(<FormGroup title={title} />);

      expect(wrapper.containsMatchingElement(title)).toBeTruthy();
    });
  });

  describe("subtitle prop", () => {
    it("renders given subtitle", () => {
      const subtitle = "Subtitle";

      const wrapper = mount(<FormGroup subtitle={subtitle} />);

      expect(wrapper.containsMatchingElement(subtitle)).toBeTruthy();
    });
  });
});
