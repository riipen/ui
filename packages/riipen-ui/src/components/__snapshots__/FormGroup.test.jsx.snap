// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormGroup> renders correct snapshot 1`] = `
<withClasses(FormGroup)>
  <FormGroup
    classes={
      Array [
        "riipen",
        "riipen-formgroup",
      ]
    }
  >
    <fieldset
      className="jsx-552593341 riipen riipen-formgroup"
    >
      <div
        className="jsx-552593341 children"
      />
    </fieldset>
    <JSXStyle
      dynamic={
        Array [
          40,
          20,
        ]
      }
      id="925181449"
    />
  </FormGroup>
</withClasses(FormGroup)>
`;
